src/main/kotlin/AVLSearchTree.kt

```diff
@@ -0,0 +1,138 @@
+/** use height for balance tree. The tree is self-balancing
+ *  because the difference between the height of the left and right subtree is -1, 0 or +1.*/
+class AVLTreeNode<K: Comparable<K>, V> (key: K, value: V) : TreeNode<K, V, AVLTreeNode<K, V>> (key, value) {
+    var height = 1
+}
+
+class AVLTreeSearch<K : Comparable<K>, V> : TreeSearch<K, V, AVLTreeNode<K, V>>() {
+    private fun insert(root: AVLTreeNode<K, V>?, key: K, value: V): AVLTreeNode<K, V> {
+        /**walk through the tree until
+         * we find the zero vertex where we will place the new element*/
+        if (root == null) return AVLTreeNode(key, value)
+        if (key < root.key) root.leftChild = insert(root.leftChild, key, value)
+        else if (key > root.key) root.rightChild = insert(root.rightChild, key, value)
+        else if (value != root.value) {
+            throw IllegalArgumentException("The key: $key already exists in the tree.")
+        } else {
+            throw IllegalArgumentException("The key: $key and value: $value already exists in the tree.")
+        }
+        updateHeight(root)
+        val balance = balanceFactor(root)
+        if (balance > 1) {
+            /**the advantage on the left,
+             * depending on whether the lower vertex is on the left or on the right,
+             * the balancing depends*/
+            if (key > root.rightChild!!.key) {
+                return rightRotate(root)
+            } else {
+                root.rightChild = leftRotate(root.rightChild!!)
+                return rightRotate(root)
+            }

```

Comment:
Проще было бы написать так, хотя ваш вариант, возможно, более понятный (то же самое можно сделать и другом случае и в remove)
```suggestion
            if (key < root.rightChild!!.key) {
                root.rightChild = leftRotate(root.rightChild!!)
            }
            return rightRotate(root)
```